todoman/model.py

```diff
@@ -231,74 +274,493 @@ def _normalize_datetime(self, x):
             x = x.replace(tzinfo=self._localtimezone)
         return x
 
+    @property
+    def list(self):
+        return self._list
 
-class Database:
+    @list.setter
+    def list(self, list_):
+        self._list = list_
+
+
+class CachedTodo:
+
+    def __init__(self, cached_data):
+        pass
+
+
+class Cache:
     """
-    This class is essentially a wrapper around a directory which contains a
-    bunch of ical files. While not a traditional SQL database, it's still *our*
-    database.
+    Caches Todos for faster read and simpler querying interface
+
+    The Cache class persists relevant[1] fields into an SQL database, which is
+    only updated if the actual file has been modified. This greatly increases
+    load times, but, more importantly, provides a simpler interface for
+    filtering/querying/sorting.
+
+    The internal sqlite database is copied fully into memory at startup, and
+    dumped again at showdown. This reduces excesive disk I/O.
+
+    [1]: Relevent fields are those we show when listing todos, or those which
+    may be used for filtering/sorting.
     """
 
     def __init__(self, path):
+        self.cache_path = path
+
+        self.conn = sqlite3.connect(
+            ':memory:',
+            detect_types=sqlite3.PARSE_DECLTYPES | sqlite3.PARSE_COLNAMES,
+        )
+        self.conn.row_factory = sqlite3.Row
+        self.load_from_disk()
+        self.cur = self.conn.cursor()
+        self.cur.execute("PRAGMA foreign_keys = ON")
+
+        self.create_tables()
+
+    def load_from_disk(self):
+        conn_disk = sqlite3.connect(
+            self.cache_path,
+            detect_types=sqlite3.PARSE_DECLTYPES | sqlite3.PARSE_COLNAMES,
+        )
+        sqlitebck.copy(conn_disk, self.conn)
```

Comment:
I was doing this for speed - to reduce disk I/O. But it seems that my initial testing was mistaken and this just made the whole thing slower. (time on my same test case without this is about 1/3 of the original one!).